harness/determined/gpu.py

```diff
@@ -119,3 +119,56 @@ def get_gpus() -> List[GPU]:
 
 def get_gpu_uuids() -> List[str]:
     return [gpu.uuid for gpu in sorted(get_gpus(), key=lambda gpu: gpu.id)]
+
+
+class GPUProcess(NamedTuple):
+    pid: int
+    process_name: str
+    gpu_uuid: str
+
+
+def _get_nvidia_processes() -> List[GPUProcess]:
+    try:
+        proc = subprocess.Popen(
+            [
+                "nvidia-smi",
+                "--query-compute-apps=" + ",".join(GPUProcess._fields),
+                "--format=csv,noheader",
+            ],
+            stdout=subprocess.PIPE,
+            universal_newlines=True,
+        )
+    except FileNotFoundError:
+        logging.info("detected 0 gpu processes (nvidia-smi not found)")
+        # This case is expected if NVIDIA drivers are not available.
+        return []
+    except Exception as e:
+        logging.warning(f"detected 0 gpu processes (error with nvidia-smi: {e})")
+        return []
+
+    processes = []
+    with proc:
+        for field_list in csv.reader(proc.stdout):  # type: ignore
+            if len(field_list) != len(GPUProcess._fields):
+                logging.warning(f"Ignoring unexpected nvidia-smi output: {field_list}")
+                continue
+            fields = dict(zip(GPUProcess._fields, field_list))
+            try:
+                processes.append(
+                    GPUProcess(
+                        pid=int(fields["pid"]),
+                        process_name=fields["process_name"],
+                        gpu_uuid=fields["gpu_uuid"].strip(),
+                    )
+                )
+            except ValueError:
+                logging.warning(f"Ignoring GPU process with unexpected nvidia-smi output: {fields}")
+    if proc.returncode:
+        logging.warning(f"nvidia-smi exited with failure status code {proc.returncode}")
+
+    logging.info(f"detected {len(processes)} gpu processes")
```

Comment:
this isn't really helpful to the user.  The only place this function is called, we already print a message when the user needs to care.  So let's delete this line.

Reply:
I thought it reasonable to include an explicit "0 processes" message at the INFO level. On second thought, though, the number of running processes will only be interesting to the user if and only if they want to be warned about a non-zero number of processes, which they'll get.
So yeah, I'll reduce the clutter.